--- _SB/Libc/string/FreeBSD/index.3	2004-11-25 11:38:46.000000000 -0800
+++ _SB/Libc/string/FreeBSD/index.3.edit	2006-06-28 16:55:53.000000000 -0700
@@ -58,8 +58,8 @@
 .Vt char )
 in the string pointed to by
 .Fa s .
-The terminating null character is considered part of the string;
-therefore if
+The terminating null character is considered to be part of the string;
+therefore, if
 .Fa c
 is
 .Ql \e0 ,
@@ -70,7 +70,7 @@
 .Fn rindex
 function is identical to
 .Fn index ,
-except it locates the last occurrence of
+except that it locates the last occurrence of
 .Fa c .
 .Sh RETURN VALUES
 The functions
