--- _SB/Libc/string/FreeBSD/rindex.3	2003-05-20 15:23:54.000000000 -0700
+++ _SB/Libc/string/FreeBSD/rindex.3.edit	2006-06-28 16:55:53.000000000 -0700
@@ -58,8 +58,8 @@
 in the null-terminated string
 .Fa s .
 .Sh RETURN VALUES
-A pointer to the character is returned if it is found; otherwise
-NULL is returned.
+A pointer to the character is returned if it is found;
+otherwise, NULL is returned.
 If
 .Fa c
 is
