--- _SB/Libc/string/FreeBSD/strcspn.3	2003-05-20 15:23:54.000000000 -0700
+++ _SB/Libc/string/FreeBSD/strcspn.3.edit	2006-06-28 16:55:53.000000000 -0700
@@ -47,27 +47,29 @@
 .Sh SYNOPSIS
 .In string.h
 .Ft size_t
-.Fn strcspn "const char *s" "const char *charset"
+.Fo strcspn
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn strcspn
 function
 spans the initial part of the null-terminated string
-.Fa s
+.Fa s1 ,
 as long as the characters from
-.Fa s
+.Fa s1
 do not occur in string
-.Fa charset
+.Fa s2
 (it
 spans the
 .Em complement
 of
-.Fa charset ) .
+.Fa s2 ) .
 .Sh RETURN VALUES
 The
 .Fn strcspn
-function
-returns the number of characters spanned.
+function returns the number of characters spanned.
 .Sh SEE ALSO
 .Xr memchr 3 ,
 .Xr strchr 3 ,
