--- _SB/Libc/string/FreeBSD/strcasecmp.3	2004-11-25 11:38:46.000000000 -0800
+++ _SB/Libc/string/FreeBSD/strcasecmp.3.edit	2006-06-28 16:55:53.000000000 -0700
@@ -39,16 +39,40 @@
 .Os
 .Sh NAME
 .Nm strcasecmp ,
-.Nm strncasecmp
+.Nm strcasecmp_l ,
+.Nm strncasecmp ,
+.Nm strncasecmp_l
 .Nd compare strings, ignoring case
 .Sh LIBRARY
 .Lb libc
 .Sh SYNOPSIS
 .In strings.h
 .Ft int
-.Fn strcasecmp "const char *s1" "const char *s2"
+.Fo strcasecmp
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fc
 .Ft int
-.Fn strncasecmp "const char *s1" "const char *s2" "size_t len"
+.Fo strncasecmp
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fa "size_t n"
+.Fc
+.In strings.h
+.In xlocale.h
+.Ft int
+.Fo strcasecmp_l
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fa "locale_t loc"
+.Fc
+.Ft int
+.Fo strncasecmp_l
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fa "size_t n"
+.Fa "locale_t loc"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn strcasecmp
@@ -63,8 +87,20 @@
 The
 .Fn strncasecmp
 compares at most
-.Fa len
+.Fa n
 characters.
+.Pp
+Although the
+.Fn strcasecmp
+and
+.Fn strncasecmp
+functions use the current locale, the
+.Fn strcasecmp_l
+and
+.Fn strncasecmp_l
+functions may be passed locales directly. See
+.Xr xlocale 3
+for more information.
 .Sh RETURN VALUES
 The
 .Fn strcasecmp
@@ -87,7 +123,8 @@
 .Xr strcmp 3 ,
 .Xr strcoll 3 ,
 .Xr strxfrm 3 ,
-.Xr tolower 3
+.Xr tolower 3 ,
+.Xr xlocale 3
 .Sh HISTORY
 The
 .Fn strcasecmp
