--- _SB/Libc/string/FreeBSD/string.3	2003-05-20 15:23:54.000000000 -0700
+++ _SB/Libc/string/FreeBSD/string.3.edit	2006-06-28 16:55:53.000000000 -0700
@@ -38,77 +38,143 @@
 .Dt STRING 3
 .Os
 .Sh NAME
+.Nm index ,
+.Nm rindex ,
 .Nm stpcpy ,
+.Nm strcasecmp ,
 .Nm strcat ,
-.Nm strncat ,
 .Nm strchr ,
-.Nm strrchr ,
 .Nm strcmp ,
-.Nm strncmp ,
-.Nm strcasecmp ,
-.Nm strncasecmp ,
 .Nm strcpy ,
-.Nm strncpy ,
+.Nm strcspn ,
 .Nm strerror ,
 .Nm strlen ,
+.Nm strncasecmp ,
+.Nm strncat ,
+.Nm strncmp ,
+.Nm strncpy ,
 .Nm strpbrk ,
+.Nm strrchr ,
 .Nm strsep ,
 .Nm strspn ,
-.Nm strcspn ,
 .Nm strstr ,
-.Nm strtok ,
-.Nm index ,
-.Nm rindex
+.Nm strtok
 .Nd string specific functions
 .Sh LIBRARY
 .Lb libc
 .Sh SYNOPSIS
-.In string.h
+.In strings.h
 .Ft char *
-.Fn stpcpy "char *dst" "const char *src"
-.Ft char *
-.Fn strcat "char *s" "const char * append"
-.Ft char *
-.Fn strncat "char *s" "const char *append" "size_t count"
-.Ft char *
-.Fn strchr "const char *s" "int c"
-.Ft char *
-.Fn strrchr "const char *s" "int c"
-.Ft int
-.Fn strcmp "const char *s1" "const char *s2"
+.Fo index
+.Fa "const char *s"
+.Fa "int c"
+.Fc
+.Ft char *
+.Fo rindex
+.Fa "const char *s"
+.Fa "int c"
+.Fc
 .Ft int
-.Fn strncmp "const char *s1" "const char *s2" "size_t count"
+.Fo strcasecmp
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fc
 .Ft int
-.Fn strcasecmp "const char *s1" "const char *s2"
-.Ft int
-.Fn strncasecmp "const char *s1" "const char *s2" "size_t count"
-.Ft char *
-.Fn strcpy "char *dst" "const char *src"
-.Ft char *
-.Fn strncpy "char *dst" "const char *src" "size_t count"
+.Fo strncasecmp
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fa "size_t n"
+.Fc
+.In string.h
 .Ft char *
-.Fn strerror "int errno"
+.Fo stpcpy
+.Fa "char *dst"
+.Fa "const char *src"
+.Fc
+.Ft char *
+.Fo strcat
+.Fa "char *restrict s1"
+.Fa "const char *restrict s2"
+.Fc
+.Ft char *
+.Fo strchr
+.Fa "const char *s"
+.Fa "int c"
+.Fc
+.Ft int
+.Fo strcmp
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fc
+.Ft char *
+.Fo strcpy
+.Fa "char *restrict s1"
+.Fa "const char *restrict s2"
+.Fc
 .Ft size_t
-.Fn strlen "const char *s"
-.Ft char *
-.Fn strpbrk "const char *s" "const char *charset"
-.Ft char *
-.Fn strsep "char **stringp" "const char *delim"
+.Fo strcspn
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fc
+.Ft char *
+.Fo strerror
+.Fa "int errnum"
+.Fc
 .Ft size_t
-.Fn strspn "const char *s" "const char *charset"
+.Fo strlen
+.Fa "const char *s"
+.Fc
+.Ft char *
+.Fo strncat
+.Fa "char *restrict s1"
+.Fa "const char *restrict s2"
+.Fa "size_t n"
+.Fc
+.Ft int
+.Fo strncmp
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fa "size_t n"
+.Fc
+.Ft char *
+.Fo strncpy
+.Fa "char *restrict s1"
+.Fa "const char *restrict s2"
+.Fa "size_t n"
+.Fc
+.Ft char *
+.Fo strpbrk
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fc
+.Ft char *
+.Fo strrchr
+.Fa "const char *s"
+.Fa "int c"
+.Fc
+.Ft char *
+.Fo strsep
+.Fa "char **stringp"
+.Fa "const char *delim"
+.Fc
 .Ft size_t
-.Fn strcspn "const char *s" "const char *charset"
-.Ft char *
-.Fn strstr "const char *big" "const char *little"
-.Ft char *
-.Fn strtok "char *s" "const char *delim"
-.Ft char *
-.Fn index "const char *s" "int c"
-.Ft char *
-.Fn rindex "const char *s" "int c"
+.Fo strspn
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fc
+.Ft char *
+.Fo strstr
+.Fa "const char *s1"
+.Fa "const char *s2"
+.Fc
+.Ft char *
+.Fo strtok
+.Fa "char *restrict s1"
+.Fa "const char *restrict s2"
+.Fc
 .Sh DESCRIPTION
 The string
-functions manipulate strings terminated by a
+functions manipulate strings that are terminated by a
 null byte.
 .Pp
 See the specific manual pages for more information.
