--- _SB/Libc/stdlib/FreeBSD/div.3	2003-05-20 15:23:24.000000000 -0700
+++ _SB/Libc/stdlib/FreeBSD/div.3.edit	2006-06-28 16:55:52.000000000 -0700
@@ -45,21 +45,23 @@
 .Sh SYNOPSIS
 .In stdlib.h
 .Ft div_t
-.Fn div "int num" "int denom"
+.Fn div "int numer" "int denom"
 .Sh DESCRIPTION
 The
 .Fn div
 function
 computes the value
-.Fa num/denom
-and returns the quotient and remainder in a structure named
+.Fa numer/denom
+(numerator/denominator).
+It returns a structure named
 .Fa div_t
 that contains two
 .Vt int
 members named
 .Va quot
-and
-.Va rem .
+(quotient) and
+.Va rem
+(remainder).
 .Sh SEE ALSO
 .Xr imaxdiv 3 ,
 .Xr ldiv 3 ,
