--- heapsort_b.c.orig	2008-09-24 13:48:45.000000000 -0700
+++ heapsort_b.c	2008-09-24 13:48:56.000000000 -0700
@@ -136,10 +136,10 @@ __FBSDID("$FreeBSD: src/lib/libc/stdlib/
  * only advantage over quicksort is that it requires little additional memory.
  */
 int
-heapsort(vbase, nmemb, size, compar)
+heapsort_b(vbase, nmemb, size, compar)
 	void *vbase;
 	size_t nmemb, size;
-	int (*compar)(const void *, const void *);
+	int (^compar)(const void *, const void *);
 {
 	size_t cnt, i, j, l;
 	char tmp, *tmp1, *tmp2;
