--- _SB/Libc/stdlib/FreeBSD/lldiv.3	2003-05-20 15:23:24.000000000 -0700
+++ _SB/Libc/stdlib/FreeBSD/lldiv.3.edit	2006-06-28 16:55:53.000000000 -0700
@@ -40,10 +40,10 @@
 The
 .Fn lldiv
 function computes the value of
-.Fa numer
+.Fa numer ,
 divided by
-.Fa denom
-and returns the stored result in the form of the
+.Fa denom .
+It returns the stored result in the form of the
 .Vt lldiv_t
 type.
 .Pp
@@ -52,7 +52,7 @@
 type is defined as:
 .Bd -literal -offset indent
 typedef struct {
-	long long quot; /* Quotient. */
+	long long quot; /* Quotient.  */
 	long long rem;  /* Remainder. */
 } lldiv_t;
 .Ed
