--- radixsort.c.orig	2004-12-01 20:08:48.000000000 -0800
+++ radixsort.c	2004-12-01 20:11:23.000000000 -0800
@@ -64,7 +64,7 @@
 } stack;
 
 static inline void simplesort
-(const u_char **, int, int, const u_char *, u_int);
+(const u_char **, int, int, const u_char *, u_int) __attribute__((always_inline));
 static void r_sort_a(const u_char **, int, int, const u_char *, u_int);
 static void r_sort_b(const u_char **, const u_char **, int, int,
     const u_char *, u_int);
