--- strtoul.3	2003-05-20 15:23:25.000000000 -0700
+++ strtoul.3.edit	2006-08-09 13:43:16.000000000 -0700
@@ -40,7 +40,10 @@
 .Dt STRTOUL 3
 .Os
 .Sh NAME
-.Nm strtoul , strtoull , strtoumax , strtouq
+.Nm strtoul ,
+.Nm strtoull ,
+.Nm strtoumax ,
+.Nm strtouq
 .Nd "convert a string to an"
 .Vt "unsigned long" , "unsigned long long" , uintmax_t ,
 or
@@ -50,25 +53,40 @@
 .Lb libc
 .Sh SYNOPSIS
 .In stdlib.h
-.In limits.h
 .Ft "unsigned long"
-.Fn strtoul "const char * restrict nptr" "char ** restrict endptr" "int base"
+.Fo strtoul
+.Fa "const char *restrict str"
+.Fa "char **restrict endptr"
+.Fa "int base"
+.Fc
 .Ft "unsigned long long"
-.Fn strtoull "const char * restrict nptr" "char ** restrict endptr" "int base"
+.Fo strtoull
+.Fa "const char *restrict str"
+.Fa "char **restrict endptr"
+.Fa "int base"
+.Fc
 .In inttypes.h
 .Ft uintmax_t
-.Fn strtoumax "const char * restrict nptr" "char ** restrict endptr" "int base"
+.Fo strtoumax
+.Fa "const char *restrict str"
+.Fa "char **restrict endptr"
+.Fa "int base"
+.Fc
 .In sys/types.h
 .In stdlib.h
 .In limits.h
 .Ft u_quad_t
-.Fn strtouq "const char *nptr" "char **endptr" "int base"
+.Fo strtouq
+.Fa "const char *str"
+.Fa "char **endptr"
+.Fa "int base"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn strtoul
 function
 converts the string in
-.Fa nptr
+.Fa str
 to an
 .Vt "unsigned long"
 value.
@@ -76,7 +94,7 @@
 .Fn strtoull
 function
 converts the string in
-.Fa nptr
+.Fa str
 to an
 .Vt "unsigned long long"
 value.
@@ -84,7 +102,7 @@
 .Fn strtoumax
 function
 converts the string in
-.Fa nptr
+.Fa str
 to an
 .Vt uintmax_t
 value.
@@ -92,7 +110,7 @@
 .Fn strtouq
 function
 converts the string in
-.Fa nptr
+.Fa str
 to a
 .Vt u_quad_t
 value.
@@ -146,11 +164,11 @@
 If there were no digits at all, however,
 .Fn strtoul
 stores the original value of
-.Fa nptr
+.Fa str
 in
 .Fa *endptr .
 (Thus, if
-.Fa *nptr
+.Fa *str
 is not
 .Ql \e0
 but
@@ -203,9 +221,21 @@
 .It Bq Er ERANGE
 The given string was out of range; the value converted has been clamped.
 .El
+.Sh LEGACY SYNOPSIS
+.Fd #include <stdlib.h>
+.Fd #include <limits.h>
+.Pp
+.In limits.h
+is necessary for the
+.Fn strtoul
+and
+.Fn strtoull
+functions.
 .Sh SEE ALSO
 .Xr strtol 3 ,
-.Xr wcstoul 3
+.Xr strtol_l 3 ,
+.Xr wcstoul 3 ,
+.Xr compat 5
 .Sh STANDARDS
 The
 .Fn strtoul
