--- daemon.c.orig	2007-09-29 23:58:54.000000000 -0700
+++ daemon.c	2007-09-30 00:46:19.000000000 -0700
@@ -37,6 +37,10 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/gen/daemon.c,v 1.6 2003/11/10 22:01:42 ghelmer Exp $");
 
+#ifndef VARIANT_PRE1050
+#include <mach/mach.h>
+#include <servers/bootstrap.h>
+#endif /* !VARIANT_PRE1050 */
 #include "namespace.h"
 #include <errno.h>
 #include <fcntl.h>
@@ -45,6 +49,33 @@
 #include <unistd.h>
 #include "un-namespace.h"
 
+#ifndef VARIANT_PRE1050
+static void
+move_to_root_bootstrap(void)
+{
+	mach_port_t parent_port = 0;
+	mach_port_t previous_port = 0;
+
+	do {
+		if (previous_port) {
+			mach_port_deallocate(mach_task_self(), previous_port);
+			previous_port = parent_port;
+		} else {
+			previous_port = bootstrap_port;
+		}
+
+		if (bootstrap_parent(previous_port, &parent_port) != 0) {
+			return;
+		}
+	} while (parent_port != previous_port);
+
+	task_set_bootstrap_port(mach_task_self(), parent_port);
+	bootstrap_port = parent_port;
+}
+#endif /* !VARIANT_PRE1050 */
+
+int daemon(int, int) __DARWIN_1050(daemon);
+
 int
 daemon(nochdir, noclose)
 	int nochdir, noclose;
@@ -60,7 +91,9 @@
 	sa.sa_handler = SIG_IGN;
 	sa.sa_flags = 0;
 	osa_ok = _sigaction(SIGHUP, &sa, &osa);
-
+#ifndef VARIANT_PRE1050
+	move_to_root_bootstrap();
+#endif /* !VARIANT_PRE1050 */
 	switch (fork()) {
 	case -1:
 		return (-1);
