--- lcong48.c.orig	2003-05-20 15:21:02.000000000 -0700
+++ lcong48.c	2005-11-03 13:36:03.000000000 -0800
@@ -16,18 +16,10 @@
 
 #include "rand48.h"
 
-extern unsigned short _rand48_seed[3];
-extern unsigned short _rand48_mult[3];
-extern unsigned short _rand48_add;
-
 void
 lcong48(unsigned short p[7])
 {
-	_rand48_seed[0] = p[0];
-	_rand48_seed[1] = p[1];
-	_rand48_seed[2] = p[2];
-	_rand48_mult[0] = p[3];
-	_rand48_mult[1] = p[4];
-	_rand48_mult[2] = p[5];
+	LOADRAND48(_rand48_seed, &p[0]);
+	LOADRAND48(_rand48_mult, &p[3]);
 	_rand48_add = p[6];
 }
