--- _SB/Libc/gen/FreeBSD/signbit.3	2003-11-12 00:44:46.000000000 -0800
+++ _SB/Libc/gen/FreeBSD/signbit.3.edit	2006-06-28 16:55:51.000000000 -0700
@@ -40,8 +40,9 @@
 The
 .Fn signbit
 macro takes an argument of
-.Fa x
-and returns non-zero if the value of its sign is negative, otherwise 0.
+.Fa x .
+It returns non-zero if the value of the argument's sign is negative,
+otherwise 0.
 .Sh SEE ALSO
 .Xr fpclassify 3 ,
 .Xr math 3
@@ -50,8 +51,3 @@
 .Fn signbit
 macro conforms to
 .St -isoC-99 .
-.Sh HISTORY
-The
-.Fn signbit
-macro was added in
-.Fx 5.1 .
