--- _SB/Libc/gen/FreeBSD/times.3	2003-05-20 15:21:03.000000000 -0700
+++ _SB/Libc/gen/FreeBSD/times.3.edit	2006-06-28 16:55:51.000000000 -0700
@@ -43,7 +43,9 @@
 .Sh SYNOPSIS
 .In sys/times.h
 .Ft clock_t
-.Fn times "struct tms *tp"
+.Fo times
+.Fa "struct tms *buffer"
+.Fc
 .Sh DESCRIPTION
 .Bf -symbolic
 This interface is obsoleted by
@@ -61,7 +63,7 @@
 Time.
 .Pp
 It also fills in the structure pointed to by
-.Fa tp
+.Fa buffer
 with time-accounting information.
 .Pp
 The
