--- usleep.c.orig	2007-05-23 18:18:17.000000000 -0700
+++ usleep.c	2007-05-23 22:15:42.000000000 -0700
@@ -31,6 +31,11 @@
  * SUCH DAMAGE.
  */
 
+#ifdef VARIANT_CANCELABLE
+#undef __DARWIN_NON_CANCELABLE
+#define __DARWIN_NON_CANCELABLE 0
+#endif /* VARIANT_CANCELABLE */
+
 #if defined(LIBC_SCCS) && !defined(lint)
 static char sccsid[] = "@(#)usleep.c	8.1 (Berkeley) 6/4/93";
 #endif /* LIBC_SCCS and not lint */
