--- _SB/Libc/gen/FreeBSD/utime.3	2003-05-20 15:21:03.000000000 -0700
+++ _SB/Libc/gen/FreeBSD/utime.3.edit	2006-06-28 16:55:51.000000000 -0700
@@ -43,7 +43,10 @@
 .Sh SYNOPSIS
 .In utime.h
 .Ft int
-.Fn utime "const char *file" "const struct utimbuf *timep"
+.Fo utime
+.Fa "const char *path"
+.Fa "const struct utimbuf *times"
+.Fc
 .Sh DESCRIPTION
 .Bf -symbolic
 This interface is obsoleted by
@@ -52,22 +55,22 @@
 .Pp
 The
 .Fn utime
-function sets the access and modification times of the named file from
-the structures in the argument array
-.Fa timep .
+function sets the access and modification times of the named file,
+based on the structures in the argument array
+.Fa times .
 .Pp
 If the times are specified (the
-.Fa timep
+.Fa times
 argument is
-.Pf non- Dv NULL )
+.Pf non- Dv NULL ) ,
 the caller must be the owner of the file or be the super-user.
 .Pp
 If the times are not specified (the
-.Fa timep
+.Fa times
 argument is
-.Dv NULL )
-the caller must be the owner of the file, have permission to write
-the file, or be the super-user.
+.Dv NULL ) ,
+the caller must be the owner of the file,
+have permission to write the file, or be the super-user.
 .Sh ERRORS
 The
 .Fn utime
