--- gb2312.c.orig	Thu Nov 25 11:38:17 2004
+++ gb2312.c	Fri Feb 18 15:45:33 2005
@@ -28,6 +28,8 @@
 #include <sys/param.h>
 __FBSDID("$FreeBSD: src/lib/libc/locale/gb2312.c,v 1.8 2004/05/12 14:09:04 tjr Exp $");
 
+#include "xlocale_private.h"
+
 #include <errno.h>
 #include <runetype.h>
 #include <stdlib.h>
@@ -35,31 +37,32 @@
 #include <wchar.h>
 #include "mblocal.h"
 
-int	_GB2312_init(_RuneLocale *);
-size_t	_GB2312_mbrtowc(wchar_t * __restrict, const char * __restrict, size_t,
-	    mbstate_t * __restrict);
-int	_GB2312_mbsinit(const mbstate_t *);
-size_t	_GB2312_wcrtomb(char * __restrict, wchar_t, mbstate_t * __restrict);
+#define GB2312_MB_CUR_MAX	2
+
+__private_extern__ int	_GB2312_init(struct __xlocale_st_runelocale *);
+static size_t	_GB2312_mbrtowc(wchar_t * __restrict, const char * __restrict, size_t,
+	    mbstate_t * __restrict, locale_t);
+static int	_GB2312_mbsinit(const mbstate_t *, locale_t);
+static size_t	_GB2312_wcrtomb(char * __restrict, wchar_t, mbstate_t * __restrict, locale_t);
 
 typedef struct {
 	int	count;
 	u_char	bytes[2];
 } _GB2312State;
 
-int
-_GB2312_init(_RuneLocale *rl)
+__private_extern__ int
+_GB2312_init(struct __xlocale_st_runelocale *xrl)
 {
 
-	_CurrentRuneLocale = rl;
-	__mbrtowc = _GB2312_mbrtowc;
-	__wcrtomb = _GB2312_wcrtomb;
-	__mbsinit = _GB2312_mbsinit;
-	__mb_cur_max = 2;
+	xrl->__mbrtowc = _GB2312_mbrtowc;
+	xrl->__wcrtomb = _GB2312_wcrtomb;
+	xrl->__mbsinit = _GB2312_mbsinit;
+	xrl->__mb_cur_max = GB2312_MB_CUR_MAX;
 	return (0);
 }
 
-int
-_GB2312_mbsinit(const mbstate_t *ps)
+static int
+_GB2312_mbsinit(const mbstate_t *ps, locale_t loc)
 {
 
 	return (ps == NULL || ((const _GB2312State *)ps)->count == 0);
@@ -88,9 +91,9 @@
 	return (1);
 }
 
-size_t
+static size_t
 _GB2312_mbrtowc(wchar_t * __restrict pwc, const char * __restrict s, size_t n,
-    mbstate_t * __restrict ps)
+    mbstate_t * __restrict ps, locale_t loc)
 {
 	_GB2312State *gs;
 	wchar_t wc;
@@ -110,7 +113,7 @@
 		pwc = NULL;
 	}
 
-	ncopy = MIN(MIN(n, MB_CUR_MAX), sizeof(gs->bytes) - gs->count);
+	ncopy = MIN(MIN(n, GB2312_MB_CUR_MAX), sizeof(gs->bytes) - gs->count);
 	memcpy(gs->bytes + gs->count, s, ncopy);
 	ocount = gs->count;
 	gs->count += ncopy;
@@ -129,8 +132,8 @@
 	return (wc == L'\0' ? 0 : len - ocount);
 }
 
-size_t
-_GB2312_wcrtomb(char * __restrict s, wchar_t wc, mbstate_t * __restrict ps)
+static size_t
+_GB2312_wcrtomb(char * __restrict s, wchar_t wc, mbstate_t * __restrict ps, locale_t loc)
 {
 	_GB2312State *gs;
 
