--- isblank.3	2004-11-25 11:38:17.000000000 -0800
+++ isblank.3.edit	2006-08-09 13:23:09.000000000 -0700
@@ -54,13 +54,13 @@
 .It "\&``\et''\t`` ''"
 .El
 .Pp
-In the "C" locale
+In the "C" locale,
 .Fn isblank
-successful test is limited to this characters only.
+successful test is limited to these characters only.
 For single C
 .Va char Ns s
 locales (see
-.Xr multibyte 3 )
+.Xr multibyte 3 ) ,
 the value of the argument is
 representable as an
 .Li unsigned char
@@ -83,6 +83,7 @@
 returns non-zero if the character tests true.
 .Sh SEE ALSO
 .Xr ctype 3 ,
+.Xr isalnum_l 3 ,
 .Xr iswblank 3 ,
 .Xr multibyte 3 ,
 .Xr ascii 7
