--- isdigit.3	2004-11-25 11:38:17.000000000 -0800
+++ isdigit.3.edit	2006-08-09 13:23:40.000000000 -0700
@@ -71,7 +71,7 @@
 For single C
 .Va char Ns s
 locales (see
-.Xr multibyte 3 )
+.Xr multibyte 3 ) ,
 the value of the argument is
 representable as an
 .Li unsigned char
@@ -83,7 +83,7 @@
 and
 .Fn isnumber
 functions return zero if the character tests false and
-returns non-zero if the character tests true.
+return non-zero if the character tests true.
 .Sh COMPATIBILITY
 The
 .Bx 4.4
@@ -96,6 +96,7 @@
 function should be used instead.
 .Sh SEE ALSO
 .Xr ctype 3 ,
+.Xr isalnum_l 3 ,
 .Xr iswdigit 3 ,
 .Xr multibyte 3 ,
 .Xr ascii 7
