--- lmonetary.h.orig	2003-05-20 15:21:44.000000000 -0700
+++ lmonetary.h	2005-02-19 14:41:58.000000000 -0800
@@ -29,6 +29,8 @@
 #ifndef _LMONETARY_H_
 #define	_LMONETARY_H_
 
+#include <xlocale.h>
+
 struct lc_monetary_T {
 	const char	*int_curr_symbol;
 	const char	*currency_symbol;
@@ -53,7 +55,7 @@
 	const char	*int_n_sign_posn;
 };
 
-struct lc_monetary_T *__get_current_monetary_locale(void);
-int	__monetary_load_locale(const char *);
+struct lc_monetary_T *__get_current_monetary_locale(locale_t);
+int	__monetary_load_locale(const char *, locale_t);
 
 #endif /* !_LMONETARY_H_ */
