--- _SB/Libc/locale/FreeBSD/localeconv.3	2004-11-25 11:38:18.000000000 -0800
+++ _SB/Libc/locale/FreeBSD/localeconv.3.edit	2006-06-29 11:04:42.000000000 -0700
@@ -40,14 +40,22 @@
 .Dt LOCALECONV 3
 .Os
 .Sh NAME
-.Nm localeconv
+.Nm localeconv ,
+.Nm localeconv_l
 .Nd natural language formatting for C
 .Sh LIBRARY
 .Lb libc
 .Sh SYNOPSIS
 .In locale.h
 .Ft struct lconv *
-.Fn localeconv "void"
+.Fo localeconv
+.Fa "void"
+.Fc
+.In xlocale.h
+.Ft struct lconv *
+.Fo localeconv_l
+.Fa "locale_t loc"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn localeconv
@@ -201,6 +209,14 @@
 A
 .Dv CHAR_MAX
 result similarly denotes an unavailable value.
+.Pp
+While the
+.Fn localeconv
+function uses the current locale, the
+.Fn localeconv_l
+function may be passed a locale directly. See
+.Xr xlocale 3
+for more information.
 .Sh RETURN VALUES
 The
 .Fn localeconv
@@ -213,7 +229,8 @@
 No errors are defined.
 .Sh SEE ALSO
 .Xr setlocale 3 ,
-.Xr strfmon 3
+.Xr strfmon 3 ,
+.Xr xlocale 3
 .Sh STANDARDS
 The
 .Fn localeconv
