--- mbsrtowcs.c.orig	2004-11-25 11:38:18.000000000 -0800
+++ mbsrtowcs.c	2005-02-18 18:28:00.000000000 -0800
@@ -27,6 +27,8 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/locale/mbsrtowcs.c,v 1.6 2004/07/21 10:54:57 tjr Exp $");
 
+#include "xlocale_private.h"
+
 #include <errno.h>
 #include <limits.h>
 #include <stdlib.h>
@@ -34,12 +36,18 @@
 #include "mblocal.h"
 
 size_t
+mbsrtowcs_l(wchar_t * __restrict dst, const char ** __restrict src, size_t len,
+    mbstate_t * __restrict ps, locale_t loc)
+{
+	NORMALIZE_LOCALE(loc);
+	if (ps == NULL)
+		ps = &loc->__mbs_mbsrtowcs;
+	return (loc->__lc_ctype->__mbsnrtowcs(dst, src, SIZE_T_MAX, len, ps, loc));
+}
+
+size_t
 mbsrtowcs(wchar_t * __restrict dst, const char ** __restrict src, size_t len,
     mbstate_t * __restrict ps)
 {
-	static mbstate_t mbs;
-
-	if (ps == NULL)
-		ps = &mbs;
-	return (__mbsnrtowcs(dst, src, SIZE_T_MAX, len, ps));
+	return mbsrtowcs_l(dst, src, len, ps, __current_locale());
 }
