--- setlocale.h.orig	2004-11-25 11:38:19.000000000 -0800
+++ setlocale.h	2005-02-19 14:45:55.000000000 -0800
@@ -29,12 +29,14 @@
 #ifndef _SETLOCALE_H_
 #define	_SETLOCALE_H_
 
+#include <xlocale.h>
+
 #define ENCODING_LEN 31
 #define CATEGORY_LEN 11
 
 extern char *_PathLocale;
 
 int	__detect_path_locale(void);
-int	__wrap_setrunelocale(const char *);
+int	__wrap_setrunelocale(const char *, locale_t);
 
 #endif /* !_SETLOCALE_H_ */
