--- _SB/Libc/locale/FreeBSD/toupper.3	2004-11-25 11:38:20.000000000 -0800
+++ _SB/Libc/locale/FreeBSD/toupper.3.edit	2006-06-28 16:55:51.000000000 -0700
@@ -40,14 +40,24 @@
 .Dt TOUPPER 3
 .Os
 .Sh NAME
-.Nm toupper
+.Nm toupper ,
+.Nm toupper_l
 .Nd lower case to upper case letter conversion
 .Sh LIBRARY
 .Lb libc
 .Sh SYNOPSIS
 .In ctype.h
 .Ft int
-.Fn toupper "int c"
+.Fo toupper
+.Fa "int c"
+.Fc
+.In ctype.h
+.In xlocale.h
+.Ft int
+.Fo toupper_l
+.Fa "int c"
+.Fa "locale_t loc"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn toupper
@@ -56,17 +66,25 @@
 For single C
 .Va char Ns s
 locales (see
-.Xr multibyte 3 )
+.Xr multibyte 3 ) ,
 the value of the argument is
 representable as an
 .Li unsigned char
 or the value of
 .Dv EOF .
+.Pp
+Although the
+.Fn toupper
+function uses the current locale, the
+.Fn toupper_l
+function may be passed a locale directly. See
+.Xr xlocale 3
+for more information.
 .Sh RETURN VALUES
 If the argument is a lower-case letter, the
 .Fn toupper
 function returns the corresponding upper-case letter if there is
-one; otherwise the argument is returned unchanged.
+one; otherwise, the argument is returned unchanged.
 .Sh COMPATIBILITY
 The
 .Bx 4.4
@@ -81,7 +99,8 @@
 .Xr ctype 3 ,
 .Xr isupper 3 ,
 .Xr multibyte 3 ,
-.Xr towupper 3
+.Xr towupper 3 ,
+.Xr xlocale 3
 .Sh STANDARDS
 The
 .Fn toupper
