--- _SB/Libc/locale/FreeBSD/towlower.3	2003-05-20 15:21:44.000000000 -0700
+++ _SB/Libc/locale/FreeBSD/towlower.3.edit	2006-06-28 16:55:51.000000000 -0700
@@ -40,29 +40,48 @@
 .Dt TOWLOWER 3
 .Os
 .Sh NAME
-.Nm towlower
+.Nm towlower ,
+.Nm towlower_l
 .Nd "upper case to lower case letter conversion (wide character version)"
 .Sh LIBRARY
 .Lb libc
 .Sh SYNOPSIS
 .In wctype.h
 .Ft wint_t
-.Fn towlower "wint_t wc"
+.Fo towlower
+.Fa "wint_t wc"
+.Fc
+.In wctype.h
+.In xlocale.h
+.Ft wint_t
+.Fo towlower_l
+.Fa "wint_t wc"
+.Fa "locale_t loc"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn towlower
 function converts an upper-case letter to the corresponding lower-case
 letter.
+.Pp
+Although the
+.Fn towlower
+function uses the current locale, the
+.Fn towlower_l
+function may be passed a locale directly. See
+.Xr xlocale 3
+for more information.
 .Sh RETURN VALUES
 If the argument is an upper-case letter, the
 .Fn towlower
 function returns the corresponding lower-case letter if there is
-one; otherwise the argument is returned unchanged.
+one; otherwise, the argument is returned unchanged.
 .Sh SEE ALSO
 .Xr iswlower 3 ,
 .Xr tolower 3 ,
 .Xr towupper 3 ,
-.Xr wctrans 3
+.Xr wctrans 3 ,
+.Xr xlocale 3
 .Sh STANDARDS
 The
 .Fn towlower
