--- _SB/Libc/locale/FreeBSD/wctrans.3	2003-05-20 15:21:45.000000000 -0700
+++ _SB/Libc/locale/FreeBSD/wctrans.3.edit	2006-06-28 16:55:51.000000000 -0700
@@ -28,21 +28,42 @@
 .Dt WCTRANS 3
 .Os
 .Sh NAME
-.Nm towctrans , wctrans
+.Nm towctrans ,
+.Nm towctrans_l ,
+.Nm wctrans ,
+.Nm wctrans_l
 .Nd "wide character mapping functions"
 .Sh LIBRARY
 .Lb libc
 .Sh SYNOPSIS
 .In wctype.h
 .Ft wint_t
-.Fn towctrans "wint_t wc" "wctrans_t desc"
+.Fo towctrans
+.Fa "wint_t wc"
+.Fa "wctrans_t desc"
+.Fc
 .Ft wctrans_t
-.Fn wctrans "const char *charclass"
+.Fo wctrans
+.Fa "const char *charclass"
+.Fc
+.In wctype.h
+.In xlocale.h
+.Ft wint_t
+.Fo towctrans_l
+.Fa "wint_t wc"
+.Fa "wctrans_t desc"
+.Fa "locale_t loc"
+.Fc
+.Ft wctrans_t
+.Fo wctrans_l
+.Fa "const char *charclass"
+.Fa "locale_t loc"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn wctrans
 function returns a value of type
-.Vt wctrans_t
+.Vt wctrans_t ,
 which represents the requested wide character mapping operation and
 may be used as the second argument for calls to
 .Fn towctrans .
@@ -55,9 +76,21 @@
 The
 .Fn towctrans
 function transliterates the wide character
-.Fa wc
+.Fa wc ,
 according to the mapping described by
 .Fa desc .
+.Pp
+While the
+.Fn towctrans
+and
+.Fn wctrans
+functions use the current locale, the
+.Fn towctrans_l
+and
+.Fn wctrans_l
+functions may be passed locales directly. See
+.Xr xlocale 3
+for more information.
 .Sh RETURN VALUES
 The
 .Fn towctrans
@@ -105,7 +138,8 @@
 .Sh SEE ALSO
 .Xr tolower 3 ,
 .Xr toupper 3 ,
-.Xr wctype 3
+.Xr wctype 3 ,
+.Xr xlocale 3
 .Sh STANDARDS
 The
 .Fn towctrans
