--- _SB/Libc/locale/FreeBSD/wcwidth.3	2004-11-25 11:38:21.000000000 -0800
+++ _SB/Libc/locale/FreeBSD/wcwidth.3.edit	2006-06-28 16:55:51.000000000 -0700
@@ -28,20 +28,38 @@
 .Dt WCWIDTH 3
 .Os
 .Sh NAME
-.Nm wcwidth
+.Nm wcwidth ,
+.Nm wcwidth_l
 .Nd "number of column positions of a wide-character code"
 .Sh LIBRARY
 .Lb libc
 .Sh SYNOPSIS
 .In wchar.h
 .Ft int
-.Fn wcwidth "wchar_t wc"
+.Fo wcwidth
+.Fa "wchar_t wc"
+.Fc
+.In wchar.h
+.In xlocale.h
+.Ft int
+.Fo wcwidth_l
+.Fa "wchar_t wc"
+.Fa "locale_t loc"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn wcwidth
 function determines the number of column positions required to
 display the wide character
 .Fa wc .
+.Pp
+Although the
+.Fn wcwidth
+function uses the current locale, the
+.Fn wcwidth_l
+function may be passed a locale directly. See
+.Xr xlocale 3
+for more information.
 .Sh RETURN VALUES
 The
 .Fn wcwidth
@@ -50,8 +68,8 @@
 argument is a null wide character (L'\e0'),
 \-1 if
 .Fa wc
-is not printable,
-otherwise it returns the number of column positions the
+is not printable;
+otherwise, it returns the number of column positions the
 character occupies.
 .Sh EXAMPLES
 This code fragment reads text from standard input and
@@ -79,7 +97,8 @@
 .Ed
 .Sh SEE ALSO
 .Xr iswprint 3 ,
-.Xr wcswidth 3
+.Xr wcswidth 3 ,
+.Xr xlocale 3
 .Sh STANDARDS
 The
 .Fn wcwidth
