--- _SB/Libc/stdio/FreeBSD/fgets.3	2003-05-20 15:22:41.000000000 -0700
+++ _SB/Libc/stdio/FreeBSD/fgets.3.edit	2006-06-28 16:55:52.000000000 -0700
@@ -48,19 +48,19 @@
 .Sh SYNOPSIS
 .In stdio.h
 .Ft char *
-.Fn fgets "char * restrict str" "int size" "FILE * restrict stream"
+.Fn fgets "char *restrict s" "int n" "FILE *restrict stream"
 .Ft char *
-.Fn gets "char *str"
+.Fn gets "char *s"
 .Sh DESCRIPTION
 The
 .Fn fgets
 function
 reads at most one less than the number of characters specified by
-.Fa size
+.Fa n
 from the given
 .Fa stream
 and stores them in the string
-.Fa str .
+.Fa s .
 Reading stops when a newline character is found,
 at end-of-file or error.
 The newline, if any, is retained.
@@ -74,7 +74,7 @@
 is equivalent to
 .Fn fgets
 with an infinite
-.Fa size
+.Fa n
 and a
 .Fa stream
 of
@@ -102,7 +102,7 @@
 and
 .Fn gets
 functions
-do not distinguish between end-of-file and error, and callers must use
+do not distinguish between end-of-file and error; callers must use
 .Xr feof 3
 and
 .Xr ferror 3
@@ -152,11 +152,6 @@
 .Xr ferror 3 ,
 .Xr fgetln 3 ,
 .Xr fgetws 3
-.Rs
-.%T "The FreeBSD Security Architecture"
-.Re
-(See
-.Pa /usr/share/doc/{to be determined} . )
 .Sh STANDARDS
 The functions
 .Fn fgets
