--- _SB/Libc/stdio/FreeBSD/fputws.3	2003-07-24 12:42:14.000000000 -0700
+++ _SB/Libc/stdio/FreeBSD/fputws.3.edit	2006-06-28 16:55:52.000000000 -0700
@@ -41,7 +41,8 @@
 .Dt FPUTWS 3
 .Os
 .Sh NAME
-.Nm fputws
+.Nm fputws ,
+.Nm fputws_l
 .Nd output a line of wide characters to a stream
 .Sh LIBRARY
 .Lb libc
@@ -49,14 +50,34 @@
 .In stdio.h
 .In wchar.h
 .Ft int
-.Fn fputws "const wchar_t * restrict ws" "FILE * restrict fp"
+.Fo fputws
+.Fa "const wchar_t *restrict ws"
+.Fa "FILE *restrict stream"
+.Fc
+.In stdio.h
+.In wchar.h
+.In xlocale.h
+.Ft int
+.Fo fputws_l
+.Fa "const wchar_t *restrict ws"
+.Fa "FILE *restrict stream"
+.Fa "locale_t loc"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn fputws
 function writes the wide character string pointed to by
 .Fa ws
 to the stream pointed to by
-.Fa fp .
+.Fa stream .
+.Pp
+Although the
+.Fn fputws
+function uses the current locale, the
+.Fn fputws_l
+function may be passed a locale directly. See
+.Xr xlocale 3
+for more information.
 .Sh RETURN VALUES
 The
 .Fn fputws
@@ -69,7 +90,7 @@
 .Bl -tag -width Er
 .It Bq Er EBADF
 The
-.Fa fp
+.Fa stream
 argument supplied
 is not a writable stream.
 .El
@@ -84,7 +105,8 @@
 .Xr ferror 3 ,
 .Xr fputs 3 ,
 .Xr putwc 3 ,
-.Xr stdio 3
+.Xr stdio 3 ,
+.Xr xlocale 3
 .Sh STANDARDS
 The
 .Fn fputws
