--- fseek.3	2004-11-25 11:38:34.000000000 -0800
+++ fseek.3.edit	2006-07-12 11:18:41.000000000 -0700
@@ -53,20 +53,39 @@
 .Sh SYNOPSIS
 .In stdio.h
 .Ft int
-.Fn fseek "FILE *stream" "long offset" "int whence"
-.Ft long
-.Fn ftell "FILE *stream"
-.Ft void
-.Fn rewind "FILE *stream"
+.Fo fgetpos
+.Fa "FILE *restrict stream"
+.Fa "fpos_t *restrict pos"
+.Fc
 .Ft int
-.Fn fgetpos "FILE * restrict stream" "fpos_t * restrict pos"
+.Fo fseek
+.Fa "FILE *stream"
+.Fa "long offset"
+.Fa "int whence"
+.Fc
 .Ft int
-.Fn fsetpos "FILE *stream" "const fpos_t *pos"
-.In sys/types.h
+.Fo fseeko
+.Fa "FILE *stream"
+.Fa "off_t offset"
+.Fa "int whence"
+.Fc
 .Ft int
-.Fn fseeko "FILE *stream" "off_t offset" "int whence"
+.Fo fsetpos
+.Fa "FILE *stream"
+.Fa "const fpos_t *pos"
+.Fc
+.Ft long
+.Fo ftell
+.Fa "FILE *stream"
+.Fc
 .Ft off_t
-.Fn ftello "FILE *stream"
+.Fo ftello
+.Fa "FILE *stream"
+.Fc
+.Ft void
+.Fo rewind
+.Fa "FILE *stream"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn fseek
@@ -246,12 +265,29 @@
 .Xr lseek 2 ,
 and
 .Xr malloc 3 .
+.Sh LEGACY SYNOPSIS
+.Fd #include <stdio.h>
+.Fd #include <sys/types.h>
+.Pp
+.Ft int
+.br
+.Fo fseeko
+.Fa "FILE *stream"
+.Fa "off_t offset"
+.Fa "int whence"
+.Fc ;
+.Pp
+The include file
+.In sys/types.h
+supplies the definition for
+.Vt off_t .
 .Sh SEE ALSO
 .Xr lseek 2 ,
 .Xr clearerr 3 ,
 .Xr fwide 3 ,
 .Xr ungetc 3 ,
-.Xr ungetwc 3
+.Xr ungetwc 3 ,
+.Xr compat 5
 .Sh STANDARDS
 The
 .Fn fgetpos ,
