--- fwscanf.c.orig	2003-05-20 15:22:42.000000000 -0700
+++ fwscanf.c	2005-02-23 16:32:29.000000000 -0800
@@ -27,6 +27,8 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/fwscanf.c,v 1.1 2002/09/23 12:40:06 tjr Exp $");
 
+#include "xlocale_private.h"
+
 #include <stdarg.h>
 #include <stdio.h>
 #include <wchar.h>
@@ -38,7 +40,21 @@
 	int r;
 
 	va_start(ap, fmt);
-	r = vfwscanf(fp, fmt, ap);
+	r = vfwscanf_l(fp, __current_locale(), fmt, ap);
+	va_end(ap);
+
+	return (r);
+}
+
+int
+fwscanf_l(FILE * __restrict fp, locale_t loc, const wchar_t * __restrict fmt, ...)
+{
+	va_list ap;
+	int r;
+
+	/* no need to call NORMALIZE_LOCALE(loc) because vfwscanf_l will */
+	va_start(ap, fmt);
+	r = vfwscanf_l(fp, loc, fmt, ap);
 	va_end(ap);
 
 	return (r);
