--- putwc.3	2004-11-25 11:38:35.000000000 -0800
+++ putwc.3.edit	2006-08-09 13:41:39.000000000 -0700
@@ -52,11 +52,19 @@
 .In stdio.h
 .In wchar.h
 .Ft wint_t
-.Fn fputwc "wchar_t wc" "FILE *stream"
+.Fo fputwc
+.Fa "wchar_t wc"
+.Fa "FILE *stream"
+.Fc
 .Ft wint_t
-.Fn putwc "wchar_t wc" "FILE *stream"
+.Fo putwc
+.Fa "wchar_t wc"
+.Fa "FILE *stream"
+.Fc
 .Ft wint_t
-.Fn putwchar "wchar_t wc"
+.Fo putwchar
+.Fa "wchar_t wc"
+.Fc
 .Sh DESCRIPTION
 The
 .Fn fputwc
@@ -79,6 +87,12 @@
 .Fn putwc
 with an output stream of
 .Dv stdout .
+.Pp
+Extended locale versions of these functions are documented in
+.Xr putwc_l 3 .
+See
+.Xr xlocale 3
+for more information.
 .Sh RETURN VALUES
 The
 .Fn fputwc ,
@@ -95,6 +109,7 @@
 .Xr fopen 3 ,
 .Xr getwc 3 ,
 .Xr putc 3 ,
+.Xr putwc_l 3 ,
 .Xr stdio 3
 .Sh STANDARDS
 The
