--- putwchar.c.orig	2004-11-25 11:38:35.000000000 -0800
+++ putwchar.c	2005-02-23 17:33:23.000000000 -0800
@@ -27,6 +27,8 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/putwchar.c,v 1.3 2004/05/25 10:42:52 tjr Exp $");
 
+#include "xlocale_private.h"
+
 #include "namespace.h"
 #include <stdio.h>
 #include <wchar.h>
@@ -43,5 +45,13 @@
 putwchar(wchar_t wc)
 {
 
-	return (fputwc(wc, stdout));
+	return (fputwc_l(wc, stdout, __current_locale()));
+}
+
+wint_t
+putwchar_l(wchar_t wc, locale_t loc)
+{
+
+	/* no need to call NORMALIZE_LOCALE(loc) because fputwc_l will */
+	return (fputwc_l(wc, stdout, loc));
 }
