Index: rewind.c
===================================================================
RCS file: /cvs/root/Libc/stdio/FreeBSD/rewind.c,v
retrieving revision 1.2
diff -u -d -b -w -p -u -r1.2 rewind.c
--- rewind.c	2003/05/20 22:22:43	1.2
+++ rewind.c	2004/10/27 05:51:34
@@ -58,8 +58,8 @@ rewind(FILE *fp)
 
 	FLOCKFILE(fp);
 	if (_fseeko(fp, (off_t)0, SEEK_SET, 1) == 0) {
-		clearerr_unlocked(fp);
 		errno = serrno;
 	}
+	clearerr_unlocked(fp);	/* POSIX: clear stdio error regardless */
 	FUNLOCKFILE(fp);
 }
