--- swscanf.c.orig	2003-05-20 15:22:44.000000000 -0700
+++ swscanf.c	2005-02-23 16:46:10.000000000 -0800
@@ -27,6 +27,8 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/swscanf.c,v 1.1 2002/09/23 12:40:06 tjr Exp $");
 
+#include "xlocale_private.h"
+
 #include <stdarg.h>
 #include <stdio.h>
 #include <wchar.h>
@@ -38,7 +40,22 @@
 	int r;
 
 	va_start(ap, fmt);
-	r = vswscanf(str, fmt, ap);
+	r = vswscanf_l(str, __current_locale(), fmt, ap);
+	va_end(ap);
+
+	return (r);
+}
+
+int
+swscanf_l(const wchar_t * __restrict str, locale_t loc,
+    const wchar_t * __restrict fmt, ...)
+{
+	va_list ap;
+	int r;
+
+	/* no need to call NORMALIZE_LOCALE(loc) because vswscanf_l will */
+	va_start(ap, fmt);
+	r = vswscanf_l(str, loc, fmt, ap);
 	va_end(ap);
 
 	return (r);
