--- vprintf.c.orig	2003-05-20 15:22:44.000000000 -0700
+++ vprintf.c	2005-02-23 16:51:20.000000000 -0800
@@ -40,11 +40,21 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/vprintf.c,v 1.10 2002/09/06 11:23:56 tjr Exp $");
 
+#include "xlocale_private.h"
+
 #include <stdio.h>
 
 int
 vprintf(const char * __restrict fmt, __va_list ap)
 {
 
-	return (vfprintf(stdout, fmt, ap));
+	return (vfprintf_l(stdout, __current_locale(), fmt, ap));
+}
+
+int
+vprintf_l(locale_t loc, const char * __restrict fmt, __va_list ap)
+{
+
+	/* no need to call NORMALIZE_LOCALE(loc) because vfprintf_l will */
+	return (vfprintf_l(stdout, loc, fmt, ap));
 }
