--- vwprintf.c.orig	2003-05-20 15:22:44.000000000 -0700
+++ vwprintf.c	2005-02-23 17:05:10.000000000 -0800
@@ -27,6 +27,8 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/vwprintf.c,v 1.1 2002/09/21 13:00:30 tjr Exp $");
 
+#include "xlocale_private.h"
+
 #include <stdarg.h>
 #include <stdio.h>
 #include <wchar.h>
@@ -35,5 +37,13 @@
 vwprintf(const wchar_t * __restrict fmt, va_list ap)
 {
 
-	return (vfwprintf(stdout, fmt, ap));
+	return (vfwprintf_l(stdout, __current_locale(), fmt, ap));
+}
+
+int
+vwprintf_l(locale_t loc, const wchar_t * __restrict fmt, va_list ap)
+{
+
+	/* no need to call NORMALIZE_LOCALE(loc) because vfwprintf_l will */
+	return (vfwprintf_l(stdout, loc, fmt, ap));
 }
