--- vwscanf.c.orig	2003-05-20 15:22:44.000000000 -0700
+++ vwscanf.c	2005-02-23 17:06:42.000000000 -0800
@@ -27,6 +27,8 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/vwscanf.c,v 1.1 2002/09/23 12:40:06 tjr Exp $");
 
+#include "xlocale_private.h"
+
 #include <stdarg.h>
 #include <stdio.h>
 #include <wchar.h>
@@ -35,5 +37,13 @@
 vwscanf(const wchar_t * __restrict fmt, va_list ap)
 {
 
-	return (vfwscanf(stdin, fmt, ap));
+	return (vfwscanf_l(stdin, __current_locale(), fmt, ap));
+}
+
+int
+vwscanf_l(locale_t loc, const wchar_t * __restrict fmt, va_list ap)
+{
+
+	/* no need to call NORMALIZE_LOCALE(loc) because vfwscanf_l will */
+	return (vfwscanf_l(stdin, loc, fmt, ap));
 }
