--- _SB/Libc/net/FreeBSD/sockatmark.3	2005-12-16 00:23:58.000000000 -0800
+++ _SB/Libc/net/FreeBSD/sockatmark.3.edit	2006-06-28 16:55:51.000000000 -0700
@@ -83,7 +83,7 @@
 .Fn sockatmark
 function returns the value 1 if the read pointer is pointing at
 the OOB mark, 0 if it is not.
-Otherwise the value \-1 is returned
+Otherwise, the value \-1 is returned
 and the global variable
 .Va errno
 is set to
